src/core/sandbox.c

```diff
@@ -0,0 +1,801 @@
+ *	- add support tame() on OpenBSD, and Capsicum on FreeBSD
+*/
+
+void create_namespaces(void)
+{
+	/* TODO */
+}
+
+void drop_privileges(void)
+{
+	int tsc_state;
+	size_t i;
+	cap_t cap;
+	cap_flag_value_t value;
+
+	/* running as root largely defeats the purpose of using a sandbox */
+	if (geteuid() == 0 || getegid() == 0) {
+		fprintf(stderr, "Cannot run irssi as root when the sandbox is enabled\n");
+		exit(1);
+	}
+
+	/* disable the rdtsc so it cannot be used for side-channel attacks */
+#ifndef IRSSI_NEEDS_RDTSC
+	if (prctl(PR_GET_TSC, &tsc_state, 0, 0, 0) < 0) {
+		fprintf(stderr, "Could not get the TSC state\n");
+		exit(1);
+	}
+
+	if (tsc_state == PR_TSC_ENABLE) {
+		prctl(PR_SET_TSC, PR_TSC_SIGSEGV, 0, 0, 0);
+		prctl(PR_GET_TSC, &tsc_state, 0, 0, 0);
+
+		if (tsc_state != PR_TSC_SIGSEGV) {
+			fprintf(stderr, "Could not disable the TSC\n");
+			exit(1);
+		}
+	}
+#endif
+
+	/* drop any dangerous capabilities the process is running with
+	 * https://forums.grsecurity.net/viewtopic.php?f=7&t=2522 */
+	/* FIXME why does cap_get_proc() fail? */
+#ifndef IRSSI_NEEDS_CAPS
+	cap = cap_get_proc();
+	if (cap == NULL) {
+		fprintf(stderr, "Could not initialize capabilities (cap_get_proc failed)\n");
+		exit(1);
+	}
+
+	if (cap_clear(cap) < 0) {
+		fprintf(stderr, "Could not clear capabilities (cap_clear failed)\n");
+		cap_free(cap);
+		exit(1);
+	}
+
+	for (i = 0; i < 64; i++) {
+		cap_get_flag(cap, i, CAP_EFFECTIVE, &value);
+		if (value == CAP_SET) {
+			fprintf(stderr, "Cannot run irssi with caps set (cap %u)\n", (unsigned)i);
+			cap_free(cap);
+			exit(1);
+		}
+	}
+
+	cap_free(cap);
+#endif
+}
+
+void enforce_resource_limits(void)
+{
+	size_t i;
+	struct rlimit rlim;
+
+	struct rlimit_struct_t {
+		int resource;
+		int limit;
+	};
+	const struct rlimit_struct_t rlimit_struct[9] = {
+		{ RLIMIT_AS,         268435456  }, /* 256 MiB */
+		{ RLIMIT_FSIZE,      1073741824 }, /* 1 GiB */
+		{ RLIMIT_LOCKS,      0          },
+		{ RLIMIT_MEMLOCK,    0          },
+		{ RLIMIT_MSGQUEUE,   0          },
+		{ RLIMIT_NOFILE,     4096       },
+		{ RLIMIT_NPROC,      70         }, /* why is this so large? */
+		{ RLIMIT_SIGPENDING, 0          },
+		{ RLIMIT_STACK,      32768      } /* 32 KiB */
+	};
+	for (i = 0; i < 9; i++) {
+		rlim.rlim_cur = rlimit_struct[i].limit;
+		rlim.rlim_max = rlimit_struct[i].limit;
+
+		if (setrlimit(rlimit_struct[i].resource, &rlim) < 0 && errno != EPERM) {
+			fprintf(stderr, "Could not set resource limits. Errno %d\n", errno);
+			exit(1);
+		}
+	}
+}
+
+void enable_seccomp_sandbox(void)
+{
+	scmp_filter_ctx ctx;
+	size_t i;
+	int rc;
+
+	/* initialize the libseccomp context */
+	ctx = seccomp_init(SCMP_ACT_KILL);
+	if (ctx == NULL) {
+		fprintf(stderr, "Could not initialize the sandbox (seccomp_init failed)\n");
+		exit(1);
+	}
+
+	/* syscalls without argument filtering */
+	const int noarg_whitelist[20] = {
+		SCMP_SYS(brk),
+		SCMP_SYS(chdir),
+		SCMP_SYS(clone),
```

Comment:
No, this is for allow (this code above verifies that the `CLONE_UNTRACED` flag is NOT set).

To check if a flag IS set, you would do `SCMP_A0(SCMP_CMP_MASKED_EQ, MY_FLAG, MY_FLAG)`